src/operation.rs

```diff
@@ -392,24 +392,57 @@ fn process_unary_operation(
                 ))
             }
         }
-        UnaryOp::MapRec(f) => {
+        UnaryOp::ListMap(f) => {
+            if let Term::List(ts) = *t {
+                let mut shared_env = Environment::new();
+                let f_as_var = f.body.closurize(&mut env, f.env);
+
+                // List elements are closurized to preserve lazyness of data structures. It
+                // maintains the invariant that any data structure only contain thunks (that is,
+                // currently, variables).
```

Comment:
For instance, I think this is not a particularly useful invariant. But, again, let's revisit later.

Reply:
I agree that this may not very important in general, but currently, this allows for simple serialization: because this invariant is enforced, when you `deepSeq` a term, you are guaranteed to only encounter values or variables. Hence the serializer can just evaluate `let x = stuff in deepSeq x x`, walk the AST of the evaluated term and serialize basic values. Here, `let x = list.map (fun x => x + 1) [1,2,3]) in deepSeq x x` would leave thunks of the form `%somevar + 1` inside, on which the serializer choked with errors, so very concretely, you couldn't export a term with a `lists.map`, which was the actual motivation behind this PRs. Sorry, I should have made this clear.